### PR TITLE
OpikEvaluationDataset test branch

### PR DESCRIPTION
**Pull request by lrcouto**
This branch is to test OpikEvaluationDataset (https://github.com/kedro-org/kedro-plugins/pull/1364). PR is to make it easier to see differences compared to main.

### How to test

- Clone the [`opik-evaluation` branch of kedro-academy](https://github.com/kedro-org/kedro-academy/tree/opik-evaluation)
- Set up credentials in `conf/local/credentials.yml`:

```yaml
opik_credentials:
  api_key: "opik-api-key"
  workspace: "workspace-name"

openai:
  openai_api_key: "openai-api-key"
```

Then we can run the pipeline and check the result on the opik dashboard:

```bash
kedro run --pipeline intent_detection_evaluation_opik --params model_name=gpt-4o
```

Expected in the Opik dashboard:
- A new dataset named `evaluations/intent_agent_evaluation` appears under Datasets
- Items are visible with auto-generated UUIDs (human-readable IDs are stripped, this is expected)
- An experiment run appears under Experiments with scoring results per item

image

Run the pipeline a second time without changing the local file. Content hash deduplication means no new items are created in Opik, the item count stays the same.

Expected: dataset item count unchanged; a new experiment run is recorded.

image

Then, add a new item to `data/intent_detection/evaluation/intent_evaluation.json` and rerun. The new item should appear in the remote dataset after load.

image

### CODE REVIEW

**Review by ElenaKhaustova — COMMENTED** [review]
Thank you, @lrcouto!

Can we please keep just the `intent_detection_evaluation_opik` pipeline here, and use `OpikEvaluationDataset` from the datasets package? I understand it is left from the prototype, but it would be easier to test it with the branch which we merge here at the end. Perhaps we can open a separate PR if you want to keep the prototype changes in this branch.

Could you please also add some e2e test for the dataset, like I've done it here:https://github.com/kedro-org/kedro-academy/blob/main/kedro-agentic-workflows/notebooks/e2e_langfuse_evaluation_dataset.ipynb

Since it has different sync policies, edge cases, etc., it is very hard to test them all manually. Having a separate file that covers these scenarios makes it much simpler for the author to double-check that everything is correct and for the reviewer. Feel free to take the same test cases I've covered.

**Comment by lrcouto** [comment]
> Thank you, @lrcouto!
> 
> Can we please keep just the `intent_detection_evaluation_opik` pipeline here, and use `OpikEvaluationDataset` from the datasets package? I understand it is left from the prototype, but it would be easier to test it with the branch which we merge here at the end. Perhaps we can open a separate PR if you want to keep the prototype changes in this branch.
> 
> Could you please also add some e2e test for the dataset, like I've done it here:https://github.com/kedro-org/kedro-academy/blob/main/kedro-agentic-workflows/notebooks/e2e_langfuse_evaluation_dataset.ipynb
> 
> Since it has different sync policies, edge cases, etc., it is very hard to test them all manually. Having a separate file that covers these scenarios makes it much simpler for the author to double-check that everything is correct and for the reviewer. Feel free to take the same test cases I've covered.

Sure, I'll adress that together with the review on the kedro-plugins PR.

**Comment by lrcouto** [comment]
Adapted the e2e test script I was using for the `evaluation_opik` pipeline (now removed) to cover the tests.

File: kedro-agentic-workflows/conf/base/evaluation_opik.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
+opik_eval_ds:
```
Comment (ElenaKhaustova) [inline kedro-agentic-workflows/conf/base/evaluation_opik.yml]: Shouldn't we have just one config file?

File: kedro-agentic-workflows/conf/base/evaluation_opik.yml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  type: kedro_agentic_workflows.datasets.opik_evaluation_dataset.OpikEvaluationDataset
```
Comment (ElenaKhaustova) [inline kedro-agentic-workflows/conf/base/evaluation_opik.yml]: Can we please change this to use it with the kedro-datasets package?

File: kedro-agentic-workflows/conf/base/evaluation_opik.yml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+opik_judge_llm:
+  type: langchain.ChatOpenAIDataset
+  kwargs:
+    model: "gpt-4o"
+    temperature: 0.0
+  credentials: openai
+
+opik_llm_judge_evaluator_prompt:
+  type: kedro_datasets_experimental.opik.OpikPromptDataset
+  filepath: data/evaluation/prompts/opik/judge_llm_prompt_v1.json  # swap to v2/v3 to test different judge wordings
```
Comment (ElenaKhaustova) [inline kedro-agentic-workflows/conf/base/evaluation_opik.yml]: For simplicity, we can stick to one version of the prompt here, cause the main purpose of the experiment is to evaluate intent detection agent.

File: kedro-agentic-workflows/conf/base/evaluation_opik.yml [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+opik_support_answer_llm:
+  type: langchain.ChatOpenAIDataset
+  kwargs:
+    model: "gpt-4o"
+    temperature: 0.0
+  credentials: openai
+
+opik_support_answer_prompt:
+  type: kedro_datasets_experimental.opik.OpikPromptDataset
+  filepath: data/evaluation/prompts/opik/support_answer_v1.json  # swap to v2/v3 to test different prompt versions
```
Comment (ElenaKhaustova) [inline kedro-agentic-workflows/conf/base/evaluation_opik.yml]: And here

File: kedro-agentic-workflows/tests/e2e_example.py [hunk 5]
```diff
@@ -0,0 +1,792 @@
+"""End-to-end scenarios for OpikEvaluationDataset — intent detection domain.
```
Comment (ElenaKhaustova) [inline kedro-agentic-workflows/tests/e2e_example.py]: Can we maybe move this file to the tests folder? Cause it's more for us than users.

File: kedro-agentic-workflows/conf/base/evaluation_opik.yml (hunk 1 above)
Comment (lrcouto) [reply kedro-agentic-workflows/conf/base/evaluation_opik.yml]: Ah this one is a leftover from that old pipeline. I'll remove it.

**Comment by lrcouto** [comment]
Made the changes, removed some other leftovers from the old pipeline that we removed.

File: kedro-agentic-workflows/notebooks/e2e_langfuse_evaluation_dataset.ipynb [hunk 6]
```diff
@@ -1288,7 +1288,7 @@
  ],
  "metadata": {
   "kernelspec": {
-   "display_name": "Python 3 (ipykernel)",
```
Comment (ElenaKhaustova) [inline kedro-agentic-workflows/notebooks/e2e_langfuse_evaluation_dataset.ipynb:1291]: Can we please discard changes for this file

File: kedro-agentic-workflows/src/kedro_agentic_workflows/settings.py [hunk 7]
```diff
@@ -32,7 +32,7 @@
     "base_env": "base",
     "default_run_env": "local",
     "config_patterns": {
-        "catalog": ["catalog*", "catalog*/**", "**/catalog*"],
+        "catalog": ["catalog*", "catalog*/**", "**/catalog*", "genai-config.yml", "evaluation_opik.yml"],
```
Comment (ElenaKhaustova) [inline kedro-agentic-workflows/src/kedro_agentic_workflows/settings.py]: Can we rename it to `catalog_evaluation_opik.yml`? Then this modification is not needed as it will be included with the default pattern.

File: kedro-agentic-workflows/src/kedro_agentic_workflows/settings.py [hunk 8]
```diff
@@ -45,9 +45,5 @@
 # DATA_CATALOG_CLASS = DataCatalog
 
 import warnings
-from kedro.utils import KedroExperimentalWarning
```
Comment (ElenaKhaustova) [inline kedro-agentic-workflows/src/kedro_agentic_workflows/settings.py:48]: Is this done intentionally?

Comment (lrcouto) [reply kedro-agentic-workflows/src/kedro_agentic_workflows/settings.py:48]: I don't think this warning is being used anywhere in the project 🤔

**Comment by lrcouto** [comment]
@ElenaKhaustova setting changes should be OK now, notebook unwanted changes also removed. I don't see any uses of KedroExperimentalWarning in this project, so I thought it was a leftover from something previous.

File: kedro-agentic-workflows/src/kedro_agentic_workflows/settings.py (hunk 8 above)
Comment (ElenaKhaustova) [reply kedro-agentic-workflows/src/kedro_agentic_workflows/settings.py:48]: It is raised by Kedro core when using the experimental feature - `llm_context_node`. If removing it, runs are getting too noisy.

File: kedro-agentic-workflows/conf/base/catalog_evaluation_opik.yml [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+opik_intent_judge_prompt:
+  type: kedro_datasets_experimental.opik.OpikPromptDataset
+  filepath: data/intent_detection/evaluation/prompts/intent_judge_llm_prompt_opik.json
+  prompt_name: "opik_intent_judge_evaluator_prompt"
+  prompt_type: "chat"
+  credentials: opik_credentials
+  sync_policy: local
+  mode: sdk
+
+# opik_client is defined in evaluation_opik.yml and shared across Opik pipelines
```
Comment (ElenaKhaustova) [inline kedro-agentic-workflows/conf/base/catalog_evaluation_opik.yml]: Shouldn't we define opik client here?

Comment (ElenaKhaustova) [reply kedro-agentic-workflows/conf/base/catalog_evaluation_opik.yml]: When I try to add it this way

```yaml
opik_client:
  type: kedro_datasets_experimental.opik.OpikTraceDataset
  credentials: opik_credentials
  mode: sdk
```

I get the following error:

```bash
DatasetError:
Failed to initialise Opik client: Opik.__init__() got an unexpected keyword argument 'endpoint'.
Failed to instantiate dataset 'opik_intent_evaluation_data' of type 'kedro_datasets_experimental.opik.opik_evaluation_dataset.OpikEvaluationDataset'.
```

**Review by ElenaKhaustova — COMMENTED** [review]
@lrcouto, could you please double-check that the project is running fine on your side after the recent changes?

I'm having some issues shared the above.

**Comment by lrcouto** [comment]
Checking it now.

File: kedro-agentic-workflows/conf/base/catalog_evaluation_opik.yml (hunk 9 above)
Comment (lrcouto) [reply kedro-agentic-workflows/conf/base/catalog_evaluation_opik.yml]: Is the kedro-datasets package installed from the main branch? OpikEvaluationDataset does not exist there yet.

Either way, I reintroduced the client. That was an error, it was on a deleted file. It ran normally here for me (it was on langchain mode on the deleted file).

File: kedro-agentic-workflows/src/kedro_agentic_workflows/pipelines/intent_detection_evaluation_opik/nodes.py [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+from ..intent_detection.agent import IntentDetectionAgent
+
+logger = logging.getLogger(__name__)
+
+
+class JudgeScore(BaseModel):
+    score: int = Field(description="Integer score between 1 and 5 inclusive.")
+
+
+def _prompt_to_template(prompt: Prompt) -> ChatPromptTemplate:
```
Comment (ElenaKhaustova) [inline kedro-agentic-workflows/src/kedro_agentic_workflows/pipelines/intent_detection_evaluation_opik/nodes.py]: Can we upload this prompt in the langchain mode straight away?

https://github.com/kedro-org/kedro-plugins/blob/be6e4dd79c7e8c2b271e5da7d4e64714c3742352/kedro-datasets/kedro_datasets_experimental/opik/opik_prompt_dataset.py#L27

Then we will not need this conversion function.

File: kedro-agentic-workflows/src/kedro_agentic_workflows/settings.py (hunk 8 above)
Comment (ElenaKhaustova) [reply kedro-agentic-workflows/src/kedro_agentic_workflows/settings.py:48]: Please revert this

File: kedro-agentic-workflows/tests/e2e_example.py (hunk 5 above)
Comment (ElenaKhaustova) [inline kedro-agentic-workflows/tests/e2e_example.py]: Can we please move this to netbook? Otherwise, it is difficult to run it step by step.

File: kedro-agentic-workflows/tests/e2e_example.py [hunk 11]
```diff
@@ -0,0 +1,792 @@
...
+
+print("\n" + "=" * 60)
+print("SCENARIO 4: save() — insert to remote + merge into local file")
+print("=" * 60)
+print("""
+Steps:
+  1. Call save() with 3 items: UUID_5 (new), UUID_6 (new),
+     UUID_1 (same UUID, updated expected_output).
+
+Expected (remote):
```
Comment (ElenaKhaustova) [inline kedro-agentic-workflows/tests/e2e_example.py]: Is this correct?

 `_UUID_1` is a valid UUID v7, and Opik upserts by ID, so the existing row should be **updated in-place**, not duplicated. The remote total should be 6 (4 existing + 2 genuinely new), not "up to 7."

Also, the test only checks the local file after `save()` — it should also verify the remote state (e.g. load via remote mode and confirm UUID_1's content was updated, UUID_5/UUID_6 exist, and no extra rows were created).

File: kedro-agentic-workflows/tests/e2e_example.py [hunk 12]
```diff
@@ -0,0 +1,792 @@
...
+Steps:
+  1. Update the UUID-id item's expected_output in the local file.
+  2. Call load() again.
+
+Expected:
+  - The UUID is still forwarded in the insert call (not stripped).
+  - Opik receives the item with the same UUID id.
+  - Remote may contain both the old and the new row (content-hash differs).
+
+Note:
```
Comment (ElenaKhaustova) [inline kedro-agentic-workflows/tests/e2e_example.py]: Outdated note about upsert semantics

File: kedro-agentic-workflows/tests/e2e_example.py [hunk 13]
```diff
@@ -0,0 +1,792 @@
...
+print("SCENARIO 9: Items without id — warning emitted, new row on every load()")
+print("=" * 60)
+print("""
+Steps:
+  1. Write 2 items with no 'id' field to a local file.
+  2. Call load() twice.
+
+Expected:
+  - Warning is logged on each load about missing/empty 'id' fields.
+  - Each load() re-inserts the items with new auto-generated UUIDs.
```
Comment (ElenaKhaustova) [inline kedro-agentic-workflows/tests/e2e_example.py]: The initial expected section says:

> Remote row count grows by 2 on the second load (no dedup possible without content-hash match — same content IS deduped; different content would grow).

But the Note right below says content-hash dedup prevents duplicates, and the assertion expects count == 2. The initial description should be rewritten to be consistent — just say unchanged content is deduplicated and the count stays at 2.

**Review by ElenaKhaustova — COMMENTED** [review]
I managed to run it, and the main scenario works as expected. However, I've noticed that the e2e tests haven't been updated to reflect the UUID processing logic. Please double-check they are up to date, cause we cannot cover them all manually.

We also need to update the `requirements.txt`

File: kedro-agentic-workflows/src/kedro_agentic_workflows/pipelines/intent_detection_evaluation_opik/nodes.py (hunk 10 above)
Comment (lrcouto) [reply kedro-agentic-workflows/src/kedro_agentic_workflows/pipelines/intent_detection_evaluation_opik/nodes.py]: The difference here is that we won't be getting the opik.Prompt object loaded from the dataset, instead we get the ChatPromptTemplate langchain object directly. That means we won't be getting the commit hash for each experiment run in the Opik dashboard. 

I think it'd be nice to keep the hash so we can use it as a filter on the dashboard, and we can update this once https://github.com/kedro-org/kedro-plugins/issues/1348 is addressed. What do you think?

File: kedro-agentic-workflows/src/kedro_agentic_workflows/settings.py [hunk 14]
```diff
@@ -47,7 +47,4 @@
 import warnings
 from kedro.utils import KedroExperimentalWarning
 
-
-# Suppresses all subsequent KedroExperimentalWarning warnings
-warnings.filterwarnings("ignore", category=KedroExperimentalWarning)
```
Comment (ElenaKhaustova) [inline kedro-agentic-workflows/src/kedro_agentic_workflows/settings.py:52]: Please keep it

File: kedro-agentic-workflows/tests/e2e_example.py [hunk 15]
```diff
@@ -0,0 +1,802 @@
+"""End-to-end scenarios for OpikEvaluationDataset — intent detection domain.
```
Comment (ElenaKhaustova) [inline kedro-agentic-workflows/tests/e2e_example.py]: WE can remove this file now, since tests are moved to the notebook

File: kedro-agentic-workflows/src/kedro_agentic_workflows/pipelines/intent_detection_evaluation_opik/nodes.py [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+            data = json.loads(data)
+        except json.JSONDecodeError:
+            return ChatPromptTemplate.from_template(data)
+    messages = [(m["role"], m["content"]) for m in data]
+    return ChatPromptTemplate.from_messages(messages)
+
+
+def init_reason_judge_evaluator(
+    intent_judge_llm: ChatOpenAI,
+    judge_llm_prompt: Prompt,
```
Comment (ElenaKhaustova) [inline kedro-agentic-workflows/src/kedro_agentic_workflows/pipelines/intent_detection_evaluation_opik/nodes.py]: This prompt now can be in `langchain` mode cause we use commit hash of `intent_prompt`, so the conversion is not needed.

File: kedro-agentic-workflows/src/kedro_agentic_workflows/pipelines/intent_detection/agent.py [hunk 17]
```diff
@@ -66,13 +69,24 @@ def __init__(self, context: LLMContext):
         self.compiled_graph: CompiledStateGraph | None = None
         self.memory: MemorySaver | None = None
 
-        # Preserve Langfuse prompt for tracing
-        # But convert to the Langchain prompt for execution
+        # Convert prompt objects to ChatPromptTemplate for execution
```
Comment (ElenaKhaustova) [inline kedro-agentic-workflows/src/kedro_agentic_workflows/pipelines/intent_detection/agent.py]: Can you please add a comment explaining why we need this

File: kedro-agentic-workflows/src/kedro_agentic_workflows/pipelines/intent_detection_evaluation_opik/nodes.py [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+    intent_evaluation_data: Dataset,
+    intent_agent_task: Callable,
+    intent_accuracy_evaluator: Callable,
+    reason_judge_evaluator: Callable,
+    intent_prompt: Prompt,
+    model_name: str,
+) -> None:
+    """Run an Opik evaluation experiment over the intent detection dataset."""
+    intent_prompt_commit = intent_prompt.commit
+    experiment_name = f"intent_eval_prompt_{intent_prompt_commit[:8]}_model_{model_name}"
```
Comment (ElenaKhaustova) [inline kedro-agentic-workflows/src/kedro_agentic_workflows/pipelines/intent_detection_evaluation_opik/nodes.py]: Can we maybe add the prompt name here?


```suggestion
    experiment_name = f"{experiment_name}_prompt_{prompt_name}_{intent_prompt_commit[:8]}_model_{model_name}"
```

**Review by ElenaKhaustova — COMMENTED** [review]
Thank you, @lrcouto!

It works as expected now. I left a few comments to further clean this PR before merging.

Could you please also update the README to include the opik part in the evaluation section?

File: kedro-agentic-workflows/src/kedro_agentic_workflows/pipelines/intent_detection/agent.py [hunk 19]
```diff
@@ -66,13 +69,29 @@ def __init__(self, context: LLMContext):
...
         self.memory: MemorySaver | None = None
 
-        # Preserve Langfuse prompt for tracing
-        # But convert to the Langchain prompt for execution
+        # The catalog can deliver intent_prompt as either a Langfuse ChatPromptClient
+        # (mode: langchain) or an Opik Prompt SDK object (mode: sdk). Both carry the
+        # same template content but expose different APIs. We normalize to
+        # ChatPromptTemplate here so the rest of the agent can use a single
+        # interface regardless of which provider is active. Langfuse prompt is also
+        # preserved for tracing purposes.
```
Comment (ElenaKhaustova) [inline kedro-agentic-workflows/src/kedro_agentic_workflows/pipelines/intent_detection/agent.py]:
```suggestion
        # The catalog can deliver intent_prompt as a Langfuse
        # (mode: langchain/sdk) or an Opik Prompt SDK object (mode: langchain/sdk). 
        # Both carry the same template content but expose different APIs.
        # We normalize to ChatPromptTemplate here, so the rest of the agent can use a single
        # interface regardless of which provider is active.

File: kedro-agentic-workflows/src/kedro_agentic_workflows/settings.py [hunk 20]
```diff
@@ -36,7 +36,7 @@
     },
 }
 
-# Class that manages Kedro's library components.
+# Class that manages Kedro's library components.Yeah
```
Comment (ElenaKhaustova) [inline kedro-agentic-workflows/src/kedro_agentic_workflows/settings.py]:
```suggestion
# Class that manages Kedro's library components.
```

**Review by ElenaKhaustova — APPROVED** [review]
Thanks! Left a few nit comments, but looks good overall 👍

We will revisit this after the dataset is released.
